### DISCUSSION TITLE
Wrong type after parse

### DISCUSSION THREAD

**Discussion by Andi365** [colinhacks/zod#3120] [Q&A] [answered]
Hi all!

I have written this schema:

```ts
export const createEmployeeAbsence = z
  .object({
    startDate: z.date(),
    endDate: z.date().optional(),
    employeeId: z.string(),
    absence: absenceReasons,
  })
    .superRefine((val, ctx) => {
      if (val.absence !== "LONGTERMSICK" && val.endDate === undefined) {
        return ctx.addIssue({
          code: z.ZodIssueCode.invalid_date,
          message: "Slutdato skal være udfyldt",
          path: ["endDate"],
        })
      }

    if (
      val.endDate &&
      isBefore(val.endDate, val.startDate) &&
      val.absence !== "LONGTERMSICK"
    ) {
      return ctx.addIssue({
        code: z.ZodIssueCode.invalid_date,
        message: "Slutdato skal være efter startdato",
        path: ["endDate"],
      })
    }
  })
  .transform((val) => {
    if (val.absence === "LONGTERMSICK") {
      return {
        ...val,
        endDate: generateUTC(new Date(2150, 0, 1)),
      }
    }

    return val
  })
  ```
  
I try to prompt the user to always provide an endDate unless absence of LONGTERMSICK is chosen, hence the first issue in the superrefine. If LONGTERMSICK is chosen, then I will transform the endDate to a date. So in every case, the endDate will be set.

When I try to parse the schema, the endDate type is set to Date | undefined. 

I dont know if this is an bug or just me?

Thank you in advance! :)

**Comment by JacobWeisenburger** [comment, marked as answer]
your code example is missing some vars so I can't fully understand what is going on, but perhaps this would be helpful:
https://github.com/colinhacks/zod#discriminated-unions

![image](https://github.com/colinhacks/zod/assets/31667350/4e11b1d9-2c2b-4670-bcee-8a5a5ace570c)
![image](https://github.com/colinhacks/zod/assets/31667350/b1747b21-9f34-48e5-b6df-06c7edd587d3)
![image](https://github.com/colinhacks/zod/assets/31667350/eaa9de20-fca9-4e23-98e1-dd44af65ea44)
